ma/maSnap.cc

```diff
@@ -857,116 +869,65 @@ long tagVertsToSnap(Adapt* a, Tag*& t)
     Vector x = getPosition(m, v);
     if (apf::areClose(s, x, 1e-12))
       continue;
-    m->setDoubleTag(v, t, &s[0]);
+    m->setDoubleTag(v, tag, &s[0]);
+    setFlag(a, v, SNAP);
     if (m->isOwned(v))
       ++n;
   }
   m->end(it);
   return m->getPCU()->Add<long>(n);
 }
 
-static void markVertsToSnap(Adapt* a, Tag* t)
+long snapTaggedVerts(Adapt* a, Tag* tag, Snapper& snapper)
 {
-  HasTag p(a->mesh, t);
-  markEntities(a, 0, p, SNAP, DONT_SNAP);
-}
+  bool shouldForce = a->input->shouldForceAdaptation;
+  a->input->shouldForceAdaptation = true; //Allows quality to decrease from snapping
 
-bool snapOneRound(Adapt* a, Tag* t, bool isSimple, long& successCount)
-{
-  markVertsToSnap(a, t);
-  if (a->mesh->hasMatching())
-    return snapMatchedVerts(a, t, isSimple, successCount);
-  else
-    return snapAllVerts(a, t, isSimple, successCount);
-}
-
-long snapTaggedVerts(Adapt* a, Tag* tag)
-{
   long successCount = 0;
-  /* there are two approaches possible here:
-   * 1- first snap all the vertices we can without any additional
-   * operation such as digging (simple snap). And then try snapping
-   * the remaining vertices that will need extra modifications (non-
-   * simple snap).
-   * 2- first do the non-simple snaps and then the simple snaps.
-   *
-   * Here we choose approach 2 for the following reasons
-   * (a) approach 2 is approximately as fast as approach 1
-   * (b) the problematic snaps will be attempted as soon as possible.
-   * This is extremely helpful because if we wait until later on
-   * bringing the vert to-be-snapped to the boundary might become more
-   * difficult due to the change in location of neighboring verticies
-   * that will be snapped before the problematic vert to-be-snapped.
-   */
-  while (snapOneRound(a, tag, false, successCount));
-  while (snapOneRound(a, tag, true, successCount));
+  bool snapped = true;
+  int prevTagged = 0;
+  while (snapped) {
+    int tagged = tagVertsToSnap(a, tag);
+    if (tagged == prevTagged) break;
+    prevTagged = tagged;
+    if (a->mesh->hasMatching())
+      snapped = snapMatchedVerts(a, tag, successCount);
+    else
+      snapped = snapAllVerts(a, tag, snapper);
+  }
+
+  a->input->shouldForceAdaptation = shouldForce;
   return successCount;
 }
 
+int collect(Adapt* a, int val) {
+  return a->mesh->getPCU()->Add<long>(val);
+}
+
 void snap(Adapt* a)
```

Comment:
Are the changes here following Li's thesis?  If so, we should add a comment here (if it isn't in the header/etc.).